text/word-rnn/word-rnn.jl

```diff
@@ -0,0 +1,139 @@
+# architecture based on https://arxiv.org/pdf/1409.2329.pdf
+include("corpus.jl")
+using Flux
+using Parameters: @with_kw
+using Statistics: mean
+using StatsBase: wsample
+
+@with_kw mutable struct Args
+    # Model Params
+    em_size::Int = 200
+    recur_size::Int = 400
+    clip::Float64 = 0.1
+    dropout::Float64 = 0.25
+
+    # Training Params
+    γ::Float64 = 2.5
+    epochs::Int = 8
+
+    # Data Params
+    nbatch::Int = 250
+    seqlen::Int = 20
+end
+
+function create_model(vocab_size, args)
+    return Chain(
+        Flux.Embedding(vocab_size, args.em_size),
+        Flux.Dropout(args.dropout),
+        Flux.LSTM(args.em_size, args.recur_size),
+        Flux.Dropout(args.dropout),
+        Flux.LSTM(args.recur_size, args.recur_size),
+        Flux.Dropout(args.dropout),
+        Flux.Dense(args.recur_size, vocab_size)
+    )
+end
+
+function train(; kws...)
+    # initialize parameter struct
+    args = Args()
+
+    # load train data and vocabulary
+    x_train, y_train, word2ind, vocab = get_data(args)
+    vocab_size = length(vocab)
+
+    # create model
+    model = create_model(vocab_size, args; kws...)
+
+    # logit cross entropy loss function
+    function loss(x, y)
+        Flux.reset!(model)
+        return mean(Flux.logitcrossentropy(model(x_i), y_i) for (x_i, y_i) in zip(x,y))
+    end
+
+    # reference to model params, and optimizer
+    ps = Flux.params(model)
+
+    # create batch iterators for data and validation
+    data_loader = zip(x_train[1:end-5], y_train[1:end-5])
+    hold_out = zip(x_train[end-5:end], y_train[end-5:end])
+
+    # used for updating hyperparameters
+    best_val_loss = nothing
+    lr = args.γ
+
+    # begin training loop
+    @info "Start Training, total $(args.epochs) epochs"
+    for epoch = 1:args.epochs
+
+        @info "Epoch $(epoch) / $(args.epochs)"
+
+        for batch in data_loader
+
+            gradient = Flux.gradient(ps) do
+                # compute loss for this batch
+                training_loss = loss(batch...)
+                return training_loss
+            end
+
+            for x in ps
+                # apply clip to handle exploding gradients
+                grad_x = clamp!(gradient[x], -args.clip, args.clip)
+                # backprop
+                x .-= lr .* grad_x
```

Comment:
In the spirit of promoting best practices, we have https://fluxml.ai/Flux.jl/stable/training/optimisers/#Flux.Optimise.ClipValue and the rest of https://fluxml.ai/Flux.jl/stable/training/optimisers/ for this. I imagine you'd want to use something more sophisticated than plain SGD anyhow.

Reply:
@ToucheSir I opted out of the library methods as this explicit calculation provided better performance. However, for the promotion of best practices I agree and I can switch to using the library methods.